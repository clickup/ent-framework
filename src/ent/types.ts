import type { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  ID,
  LoadByInput,
  Order,
  Table,
  UniqueKey,
  Where,
} from "../types";
import type { Validation } from "./Validation";
import type { VC } from "./VC";

/**
 * A very shallow interface of Ent class (as a collection of static methods).
 * Used in some places where we need the very minimum from the Ent.
 */
export interface EntClass<TTable extends Table = any> {
  readonly SCHEMA: Schema<TTable>;
  readonly VALIDATION: Validation<TTable>;
  readonly name: string; // class constructor name

  new (): Ent;
  loadX(vc: VC, id: string): Promise<Ent>;
  loadNullable(vc: VC, id: string): Promise<Ent | null>;
  loadIfReadableNullable(vc: VC, id: string): Promise<Ent | null>;
  count(vc: VC, where: CountInput<TTable>): Promise<number>;
  exists(vc: VC, where: ExistsInput<TTable>): Promise<boolean>;
  select(
    vc: VC,
    where: Where<TTable>,
    limit: number,
    order?: Order<TTable>
  ): Promise<Ent[]>;
  selectChunked(
    vc: VC,
    where: Where<TTable>,
    chunkSize: number,
    limit: number,
    custom?: {}
  ): AsyncIterableIterator<Ent[]>;
  loadByX(vc: VC, keys: LoadByInput<TTable, UniqueKey<TTable>>): Promise<Ent>;
}

/**
 * A very shallow interface of one Ent.
 */
export interface Ent {
  readonly [ID]: string;
  readonly vc: VC;
  deleteOriginal(): Promise<boolean>;
}
