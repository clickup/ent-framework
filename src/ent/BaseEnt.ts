import type { Client } from "../abstract/Client";
import type { Cluster } from "../abstract/Cluster";
import type { Schema } from "../abstract/Schema";
import type { Table, UniqueKey } from "../types";
import { CacheMixin } from "./mixins/CacheMixin";
import { ConfigMixin } from "./mixins/ConfigMixin";
import type { HelpersClass } from "./mixins/HelpersMixin";
import { HelpersMixin } from "./mixins/HelpersMixin";
import { PrimitiveMixin } from "./mixins/PrimitiveMixin";

/**
 * This is a helper function to create new Ent classes. Run once per each
 * Ent+Cluster on app boot. See examples in __tests__/TestObjects.ts and
 * EntTest.ts.
 *
 * Since all Ent objects are immutable (following the modern practices),
 * 1. Ent is not a DataMapper pattern;
 * 2. Ent is not an ActiveRecord;
 * 3. At last, Ent is not quite a DAO pattern too.
 *
 * We assume that Ents are very simple (we don't need triggers or multi-ent
 * touching mutations), because we anyway have a GraphQL layer on top of it.
 *
 * Finally, a naming decision has been made: we translate database field names
 * directly to Ent field names, no camelCase. This has proven its simplicity
 * benefits in the past: the less translation layers we have, the easier it is
 * to debug and develop.
 */
export function BaseEnt<
  TTable extends Table,
  TUniqueKey extends UniqueKey<TTable>,
  TClient extends Client,
>(
  cluster: Cluster<TClient>,
  schema: Schema<TTable, TUniqueKey>,
): HelpersClass<TTable, TUniqueKey, TClient> {
  return HelpersMixin(
    CacheMixin(PrimitiveMixin(ConfigMixin(class {}, cluster, schema))),
  );
}
