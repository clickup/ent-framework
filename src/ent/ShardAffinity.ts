import type { ID } from "../types";

/**
 * The table is located in the global Shard (0).
 */
export const GLOBAL_SHARD = "global_shard";

/**
 * Defines Ent Shard collocation to some Ent's field when this Ent is inserted.
 * - The Shard can always be Shard 0 ("global Shard"), be inferred based on the
 *   value in other Ent field during the insertion ("colocation"), or, in case
 *   colocation inference didn't succeed, be chosen pseudo-randomly at insertion
 *   time ("random Shard").
 * - E.g. a random Shard can also be chosen in case an empty array is passed to
 *   Shard affinity (like "always fallback"), or when a field's value points to
 *   a global Shard.
 * - Passing ID to ShardAffinity is prohibited by TS.
 */
export type ShardAffinity<
  TField extends string,
  TF = Exclude<TField, typeof ID>,
> =
  | typeof GLOBAL_SHARD
  | (TField extends typeof ID ? readonly TF[] : readonly [TF, ...TF[]]);
