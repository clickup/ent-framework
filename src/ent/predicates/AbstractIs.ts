import type { Predicate } from "./Predicate";

/**
 * A base interface for all user validation predicates.
 */
export interface AbstractIs<TRow> extends Predicate<TRow> {
  /** The field this validation predicate is related to (null means that it
   * applies to the entire Ent). */
  readonly field: string | null;
  /** In case the predicate returns false or doesn't provide error messages by
   * throwing EntValidationError, this message will be used. When message is
   * null, it means that we expect the validator to return detailed information
   * about each field errored (e.g. ValidatorStandardSchemaResult).  */
  readonly message: string | null;
}

/**
 * Result of plain validators.
 */
export type ValidatorPlainResult = boolean;

/**
 * Result of Zod safeParse() calls.
 */
export type ValidatorZodSafeParseResult =
  | {
      readonly success: true;
      readonly error?: undefined;
    }
  | {
      readonly success: false;
      readonly error: {
        readonly issues: ReadonlyArray<{
          readonly message: string;
          readonly path: readonly PropertyKey[];
        }>;
      };
    };

/**
 * Result of a Standard Schema validators: https://standardschema.dev/
 */
export type ValidatorStandardSchemaResult =
  | {
      readonly value: unknown;
      readonly issues?: undefined;
    }
  | {
      readonly issues: Array<{
        readonly message: string;
        readonly path?:
          | ReadonlyArray<PropertyKey | { readonly key: PropertyKey }>
          | undefined;
      }>;
    };
