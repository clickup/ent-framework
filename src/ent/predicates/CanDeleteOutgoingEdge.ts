import type { EntClass } from "../types";
import type { VC } from "../VC";
import type { Predicate } from "./Predicate";
import { IDsCacheDeletable } from "./Predicate";

/**
 * Checks that an Ent available via a field can be deleted, or Ent doesn't exist
 * (e.g. Ent is orphaned). See CanReadOutgoingEdge comments for more details.
 */
export class CanDeleteOutgoingEdge<TField extends string>
  implements Predicate<Record<TField, string | null>>
{
  readonly name;

  constructor(
    public readonly field: TField,
    public readonly toEntClass: EntClass,
  ) {
    this.name = this.constructor.name + "(" + this.field + ")";
  }

  async check(vc: VC, row: Record<TField, string | null>): Promise<boolean> {
    const toID = row[this.field];
    if (!toID) {
      return false;
    }

    const cache = vc.cache(IDsCacheDeletable);
    if (cache.has(toID!)) {
      return true;
    }

    // Load the target Ent and check that it's deletable.
    const toEnt = await this.toEntClass.loadNullable(vc, toID!);
    if (toEnt === null) {
      return true;
    }

    await this.toEntClass.VALIDATION.validateDelete(vc, toEnt);

    // Sill here and not thrown? save to the cache.
    cache.add(toID!);
    return true;
  }
}
