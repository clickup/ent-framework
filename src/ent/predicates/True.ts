import { VC } from "../VC";
import { Predicate } from "./Predicate";

/**
 * Always passes; used for e.g. globally accessed objects.
 */
export class True implements Predicate<never> {
  readonly name = this.constructor.name;

  async check(_vc: VC, _row: object): Promise<boolean> {
    return true;
  }
}
