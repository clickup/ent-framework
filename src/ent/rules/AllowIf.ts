import { EntAccessError } from "../errors/EntAccessError";
import type { VC } from "../VC";
import type { RuleResult } from "./Rule";
import { Rule } from "./Rule";

/**
 * Returns ALLOW if the predicate succeeds, otherwise SKIP.
 * - Used mostly for read permission checks.
 * - This rule may still throw an exception if the exception is a wild one (not
 *   derived from EntAccessError).
 */
export class AllowIf<TInput extends object> extends Rule<TInput> {
  readonly _TAG!: "AllowIf";

  async evaluate(vc: VC, input: TInput): Promise<RuleResult> {
    try {
      return (await this.predicate.check(vc, input))
        ? { decision: "ALLOW", rule: this, cause: null }
        : { decision: "SKIP", rule: this, cause: null };
    } catch (error: unknown) {
      if (error instanceof EntAccessError) {
        // We carry a cause for this SKIP decision too if it was due to an
        // access-related error.
        return { decision: "SKIP", rule: this, cause: error };
      }

      throw error;
    }
  }
}
