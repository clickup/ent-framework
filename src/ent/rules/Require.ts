import type { VC } from "../VC";
import type { RuleResult } from "./Rule";
import { Rule, RuleDecision } from "./Rule";

/**
 * Returns TOLERATE if the predicate succeeds, otherwise DENY.
 * - Used mostly for write permission checks.
 * - This rule may still throw an exception if it's a wild one (i.e. not derived
 *   from EntAccessError).
 */
export class Require<TInput extends object> extends Rule<TInput> {
  readonly _TAG!: "Require";

  async evaluate(vc: VC, input: TInput): Promise<RuleResult> {
    return (await this.predicate.check(vc, input))
      ? { decision: RuleDecision.TOLERATE, rule: this, cause: null }
      : { decision: RuleDecision.DENY, rule: this, cause: null };
  }
}
