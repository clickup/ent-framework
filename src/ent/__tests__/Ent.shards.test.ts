import { MASTER } from "../../abstract/Shard";
import { recreateTestTables, testCluster } from "../../pg/__tests__/test-utils";
import { PgSchema } from "../../pg/PgSchema";
import { ID } from "../../types";
import { BaseEnt } from "../BaseEnt";
import { True } from "../predicates/True";
import { AllowIf } from "../rules/AllowIf";
import { Require } from "../rules/Require";
import { GLOBAL_SHARD } from "../ShardAffinity";
import { createVC } from "./test-utils";

const ID_FROM_EXISTING_SHARD = "500010000000";
const ID_FROM_UNKNOWN_SHARD = "511110000000";
const PARENT_ID_FROM_UNKNOWN_SHARD = "511117654321";
const UNIVERSE_FROM_UNKNOWN_SHARD = "511119999999";

const vc = createVC();

class EntTestUniverse extends BaseEnt(
  testCluster,
  new PgSchema(
    'ent.shards"universe',
    {
      id: { type: ID, autoInsert: "id_gen()" },
    },
    [],
  ),
) {
  static readonly CREATE = [
    `CREATE TABLE %T(
      id bigint NOT NULL
    )`,
  ];

  static override configure() {
    return new this.Configuration({
      shardAffinity: GLOBAL_SHARD,
      privacyInferPrincipal: null,
      privacyLoad: [new AllowIf(new True())],
      privacyInsert: [new Require(new True())],
    });
  }
}

class EntTestHuman extends BaseEnt(
  testCluster,
  new PgSchema(
    'ent.shards"human',
    {
      id: { type: ID, autoInsert: "id_gen()" },
      parent_id: { type: ID },
      universe_id: { type: ID, allowNull: true },
      name: { type: String },
    },
    [],
  ),
) {
  static readonly CREATE = [
    `CREATE TABLE %T(
      id bigint NOT NULL,
      parent_id bigint NOT NULL,
      universe_id bigint,
      name text
    )`,
  ];

  static override configure() {
    return new this.Configuration({
      shardAffinity: ["parent_id"],
      inverses: {
        universe_id: { name: "inverses", type: "universe2humans" },
      },
      privacyInferPrincipal: null,
      privacyLoad: [new AllowIf(new True())],
      privacyInsert: [new Require(new True())],
    });
  }
}

beforeEach(async () => {
  await recreateTestTables([EntTestUniverse, EntTestHuman]);

  testCluster.options.locateIslandErrorRetryCount = 2;
  testCluster.options.locateIslandErrorRediscoverClusterDelayMs = 1000;
});

test("unknown shard failure, singleShardFromID", async () => {
  await expect(
    EntTestHuman.loadNullable(vc, ID_FROM_UNKNOWN_SHARD),
  ).rejects.toThrowErrorMatchingSnapshot();
  expect(testCluster.options.loggers.locateIslandErrorLogger).toBeCalledTimes(
    1,
  ); // not retried on intent
});

test("unknown shard failure, singleShardForInsert", async () => {
  await expect(
    EntTestHuman.insertIfNotExists(vc, {
      parent_id: PARENT_ID_FROM_UNKNOWN_SHARD,
      universe_id: null,
      name: "test",
    }),
  ).rejects.toThrowErrorMatchingSnapshot();
  expect(testCluster.options.loggers.locateIslandErrorLogger).toBeCalledTimes(
    1,
  ); // not retried on intent
});

test("unknown shard failure, multiShardsFromInput, no inverses", async () => {
  await expect(
    EntTestHuman.select(vc, { universe_id: UNIVERSE_FROM_UNKNOWN_SHARD }, 1),
  ).rejects.toThrowErrorMatchingSnapshot();
});

test("shard relocation error when accessing a table should be retried", async () => {
  testCluster.options.locateIslandErrorRediscoverClusterDelayMs = 1;

  const master = await testCluster.shard(ID_FROM_EXISTING_SHARD).client(MASTER);
  await master.rows("DROP TABLE %T CASCADE", EntTestHuman.SCHEMA.name);

  await expect(
    EntTestHuman.loadNullable(vc, ID_FROM_EXISTING_SHARD),
  ).rejects.toThrow(/undefined_table/);
  expect(testCluster.options.loggers.locateIslandErrorLogger).toBeCalledTimes(
    3,
  );
});
