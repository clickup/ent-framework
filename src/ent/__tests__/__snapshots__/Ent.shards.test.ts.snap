// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown shard failure, multiShardsFromInput, no inverses 1`] = `
"EntTestHuman not found: universe_id=511119999999, because:
  ShardError: Shard 1111 is not discoverable (no such Shard in the Cluster? some Islands are down? connections limit?)"
`;

exports[`unknown shard failure, singleShardForInsert 1`] = `
"EntTestHuman not found: parent_id=511117654321, because:
  ShardError: Shard 1111 is not discoverable (no such Shard in the Cluster? some Islands are down? connections limit?)"
`;

exports[`unknown shard failure, singleShardFromID 1`] = `
"EntTestHuman not found: id=511110000000, because:
  ShardError: Shard 1111 is not discoverable (no such Shard in the Cluster? some Islands are down? connections limit?)"
`;
